File: examples/coding_assistant.go
// Coding Assistant Example (Go version)
// Demonstrates using Local LLM for code generation and review
package main

import (
	"bufio"
	"context"
	"fmt"
	"os"
	"strings"
	"time"

	llmclient "github.com/mustzee/serve/clients/go"
)

// CodingAssistant provides AI-powered coding assistance
type CodingAssistant struct {
	client *llmclient.Client
	model  string
}

// NewCodingAssistant creates a new coding assistant
func NewCodingAssistant(baseURL, model string) *CodingAssistant {
	if model == "" {
		model = "codellama"
	}

	return &CodingAssistant{
		client: llmclient.NewClient(baseURL),
		model:  model,
	}
}

// GenerateCode generates code based on a prompt
func (ca *CodingAssistant) GenerateCode(ctx context.Context, prompt, language string) (string, error) {
	req := &llmclient.ChatCompletionRequest{
		Model: ca.model,
		Messages: []llmclient.Message{
			{
				Role: "system",
				Content: fmt.Sprintf("You are an expert %s programmer. Generate clean, "+
					"efficient, and well-documented code.", language),
			},
			{
				Role:    "user",
				Content: fmt.Sprintf("Generate %s code for: %s", language, prompt),
			},
		},
		Temperature: 0.3,
		MaxTokens:   2048,
	}

	resp, err := ca.client.Chat(ctx, req)
	if err != nil {
		return "", err
	}

	if len(resp.Choices) == 0 {
		return "", fmt.Errorf("no response from model")
	}

	return resp.Choices[0].Message.Content, nil
}

// ReviewCode reviews code and provides suggestions
func (ca *CodingAssistant) ReviewCode(ctx context.Context, code, language string) (string, error) {
	req := &llmclient.ChatCompletionRequest{
		Model: ca.model,
		Messages: []llmclient.Message{
			{
				Role: "system",
				Content: "You are a senior code reviewer. Provide constructive feedback " +
					"on code quality, potential bugs, performance, and best practices.",
			},
			{
				Role: "user",
				Content: fmt.Sprintf("Review this %s code:\n\n```%s\n%s\n```",
					language, language, code),
			},
		},
		Temperature: 0.5,
		MaxTokens:   2048,
	}

	resp, err := ca.client.Chat(ctx, req)
	if err != nil {
		return "", err
	}

	if len(resp.Choices) == 0 {
		return "", fmt.Errorf("no response from model")
	}

	return resp.Choices[0].Message.Content, nil
}

// ChatStream performs interactive chat with streaming
func (ca *CodingAssistant) ChatStream(ctx context.Context, prompt string) error {
	req := &llmclient.ChatCompletionRequest{
		Model: ca.model,
		Messages: []llmclient.Message{
			{
				Role: "system",
				Content: "You are a helpful coding assistant for Python and Go.",
			},
			{
				Role:    "user",
				Content: prompt,
			},
		},
		Temperature: 0.7,
		MaxTokens:   2048,
		Stream:      true,
	}

	fmt.Print("Assistant: ")
	contentChan, errorChan := ca.client.StreamChat(ctx, req)

	for {
		select {
		case content, ok := <-contentChan:
			if !ok {
				fmt.Println()
				return nil
			}
			fmt.Print(content)
		case err := <-errorChan:
			if err != nil {
				return err
			}
		case <-ctx.Done():
			return ctx.Err()
		}
	}
}

func main() {
	fmt.Println("=== Coding Assistant Demo (Go) ===\n")

	assistant := NewCodingAssistant("http://localhost:8000", "codellama")
	ctx := context.Background()

	// Check server health
	health, err := assistant.client.Health(ctx)
	if err != nil {
		fmt.Printf("✗ Server not available: %v\n", err)
		fmt.Println("Please start the server first")
		return
	}

	fmt.Printf("✓ Server Status: %s\n", health.Status)
	fmt.Printf("✓ Backend: %s\n", health.Backend)
	fmt.Printf("✓ Models: %s\n\n", strings.Join(health.ModelsLoaded, ", "))

	// Example 1: Generate code
	fmt.Println("1. Generate Code")
	fmt.Println(strings.Repeat("-", 50))
	prompt := "Create a function to implement binary search in Go"
	fmt.Printf("Prompt: %s\n\n", prompt)

	code, err := assistant.GenerateCode(ctx, prompt, "go")
	if err != nil {
		fmt.Printf("Error: %v\n", err)
	} else {
		fmt.Printf("Generated Code:\n%s\n", code)
	}

	// Example 2: Review code
	fmt.Println("\n2. Review Code")
	fmt.Println(strings.Repeat("-", 50))
	sampleCode := `
func sum(nums []int) int {
    total := 0
    for i := 0; i < len(nums); i++ {
        total = total + nums[i]
    }
    return total
}
`
	fmt.Printf("Code to review:%s\n", sampleCode)

	review, err := assistant.ReviewCode(ctx, strings.TrimSpace(sampleCode), "go")
	if err != nil {
		fmt.Printf("Error: %v\n", err)
	} else {
		fmt.Printf("Review:\n%s\n", review)
	}

	// Example 3: Interactive chat
	fmt.Println("\n3. Interactive Chat (type 'exit' to quit)")
	fmt.Println(strings.Repeat("-", 50))

	scanner := bufio.NewScanner(os.Stdin)
	for {
		fmt.Print("\nYou: ")
		if !scanner.Scan() {
			break
		}

		input := strings.TrimSpace(scanner.Text())
		if input == "" {
			continue
		}

		if strings.ToLower(input) == "exit" || strings.ToLower(input) == "quit" {
			fmt.Println("Goodbye!")
			break
		}

		ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
		if err := assistant.ChatStream(ctx, input); err != nil {
			fmt.Printf("Error: %v\n", err)
		}
		cancel()
	}
}

File: examples/coding_assistant.py
#!/usr/bin/env python3
"""
Coding Assistant Example
Demonstrates using Local LLM for code generation, review, and debugging
"""
import sys
import os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..'))

from clients.python.client import LLMClient, Message


class CodingAssistant:
    """AI-powered coding assistant"""

    def __init__(self, base_url: str = "http://localhost:8000", model: str = "codellama"):
        self.client = LLMClient(base_url)
        self.model = model
        self.conversation_history = []

    def generate_code(self, prompt: str, language: str = "python") -> str:
        """Generate code based on prompt"""
        system_message = Message(
            role="system",
            content=f"You are an expert {language} programmer. Generate clean, "
                    f"efficient, and well-documented code."
        )

        user_message = Message(
            role="user",
            content=f"Generate {language} code for: {prompt}"
        )

        messages = [system_message, user_message]
        response = self.client.chat(messages=messages, model=self.model, temperature=0.3)

        return response['choices'][0]['message']['content']

    def review_code(self, code: str, language: str = "python") -> str:
        """Review code and provide suggestions"""
        system_message = Message(
            role="system",
            content="You are a senior code reviewer. Provide constructive feedback "
                    "on code quality, potential bugs, performance, and best practices."
        )

        user_message = Message(
            role="user",
            content=f"Review this {language} code:\n\n```{language}\n{code}\n```"
        )

        messages = [system_message, user_message]
        response = self.client.chat(messages=messages, model=self.model, temperature=0.5)

        return response['choices'][0]['message']['content']

    def debug_code(self, code: str, error: str, language: str = "python") -> str:
        """Help debug code with errors"""
        system_message = Message(
            role="system",
            content="You are an expert debugger. Analyze code errors and provide "
                    "clear explanations and fixes."
        )

        user_message = Message(
            role="user",
            content=f"Debug this {language} code:\n\n"
                    f"Code:\n```{language}\n{code}\n```\n\n"
                    f"Error:\n```\n{error}\n```"
        )

        messages = [system_message, user_message]
        response = self.client.chat(messages=messages, model=self.model, temperature=0.3)

        return response['choices'][0]['message']['content']

    def explain_code(self, code: str, language: str = "python") -> str:
        """Explain what code does"""
        system_message = Message(
            role="system",
            content="You are a technical educator. Explain code clearly and concisely."
        )

        user_message = Message(
            role="user",
            content=f"Explain this {language} code:\n\n```{language}\n{code}\n```"
        )

        messages = [system_message, user_message]
        response = self.client.chat(messages=messages, model=self.model, temperature=0.5)

        return response['choices'][0]['message']['content']

    def chat_stream(self, user_input: str):
        """Interactive chat with streaming"""
        self.conversation_history.append(Message(role="user", content=user_input))

        print("Assistant: ", end="", flush=True)
        full_response = ""

        for chunk in self.client.chat(
            messages=self.conversation_history,
            model=self.model,
            stream=True
        ):
            print(chunk, end="", flush=True)
            full_response += chunk

        print()  # New line after streaming
        self.conversation_history.append(
            Message(role="assistant", content=full_response)
        )


def main():
    """Demo the coding assistant"""
    print("=== Coding Assistant Demo ===\n")

    assistant = CodingAssistant()

    # Check if server is available
    try:
        health = assistant.client.health()
        print(f"✓ Server Status: {health['status']}")
        print(f"✓ Backend: {health['backend']}")
        print(f"✓ Models: {', '.join(health['models_loaded'])}\n")
    except Exception as e:
        print(f"✗ Server not available: {e}")
        print("Please start the server first: python -m server.main")
        return

    # Example 1: Generate code
    print("1. Generate Code")
    print("-" * 50)
    prompt = "Create a function to calculate fibonacci numbers with memoization"
    print(f"Prompt: {prompt}\n")
    code = assistant.generate_code(prompt, "python")
    print(f"Generated Code:\n{code}\n")

    # Example 2: Review code
    print("\n2. Review Code")
    print("-" * 50)
    sample_code = """
def sum_list(lst):
    total = 0
    for i in range(len(lst)):
        total = total + lst[i]
    return total
"""
    print(f"Code to review:{sample_code}")
    review = assistant.review_code(sample_code.strip(), "python")
    print(f"Review:\n{review}\n")

    # Example 3: Interactive chat
    print("\n3. Interactive Chat (type 'exit' to quit)")
    print("-" * 50)

    assistant.conversation_history = [
        Message(
            role="system",
            content="You are a helpful coding assistant for Python and Go."
        )
    ]

    while True:
        try:
            user_input = input("\nYou: ").strip()
            if user_input.lower() in ['exit', 'quit', 'bye']:
                print("Goodbye!")
                break

            if not user_input:
                continue

            assistant.chat_stream(user_input)

        except KeyboardInterrupt:
            print("\n\nGoodbye!")
            break
        except Exception as e:
            print(f"Error: {e}")


if __name__ == "__main__":
    main()
